proxy/userspace: respect minSyncInterval

dcbw opened the pull request:
The userspace proxy does not have any ratelimiting and when many
services are used will hammer iptables every time a service or
endpoint change occurs.

https://bugzilla.redhat.com/show_bug.cgi?id=1590589

/kind bug
/sig network

```release-note
The userspace proxy now respects the IPTables proxy's minSyncInterval parameter.
```

dcbw committed:
proxy: consolidate ServicesHandler/EndpointsHandler into ProxyProvider

Proxies should be able to cleanly figure out when endpoints have been synced,
so make all ProxyProviders also implement EndpointsHandler and pass those
through to loadbalancers when required.

dcbw commented:
/priority important-soon

dcbw commented:
/assign @thockin

dcbw commented:
/test pull-kubernetes-e2e-kops-aws

dcbw commented:
/test pull-kubernetes-kubemark-e2e-gce-big

danwinship commented:
/lgtm

dcbw commented:
@thockin @bowei any thoughts on this one after the holidays?

thockin commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -216,7 +237,21 @@ func createProxier(loadBalancer LoadBalancer, listenIP net.IP, iptables iptables
+	proxier.syncRunner = async.NewBoundedFrequencyRunner("sync-runner", proxier.syncProxyRules, minSyncPeriod, syncPeriod, burstSyncs)
+
+	go func() {
+		<-loadBalancer.EndpointsSyncedChan()
```
Why does this only wait for endpoints and not services?  If that's intentional, it could be commented.

dcbw replied:
@thockin services are waited for via OnServicesSynced(), but because the userspace Proxier doesn't implement the EndpointsHandler interface (instead the RoundRobin LB does) we have to use a different way to figure out if endpoints are synced hence the passthrough.

The iptables proxier implements both ServicesHandler and EndpointsHandler so it doesn't have this issue.

dcbw replied:
@thockin re-implemented this so that the proxy now implements EndpointsHandler and can intercept the OnEndpointsSynced().

thockin commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -289,6 +324,44 @@ func CleanupLeftovers(ipt iptables.Interface) (encounteredError bool) {
+
+func (proxier *Proxier) syncProxyRules() {
+	start := time.Now()
+	defer func() {
```
maybe this should go after the fast-return if not initialized?

dcbw replied:
yep

thockin commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -510,20 +577,41 @@ func (proxier *Proxier) unmergeService(service *v1.Service, existingPorts sets.S
+	}
+
+	proxier.mu.Lock()
+	proxier.serviceEvents = append(proxier.serviceEvents, e)
```
If there are a lot of events (enough to cause problems syncing), won't this balloon memory usage as you save copies of every Service and Endpoints snapshot?

dcbw replied:
@thockin it only copies Services; but added a max queue depth that will trigger an immediate sync regardless of the ratelimiting.

thockin commented on pkg/proxy/userspace/roundrobin.go:
```diff
@@ -356,6 +358,11 @@ func (lb *LoadBalancerRR) OnEndpointsDelete(endpoints *v1.Endpoints) {
 }
 
 func (lb *LoadBalancerRR) OnEndpointsSynced() {
+	close(lb.endpointsSynced)
```
Why do we have 2 different mechanisms for being notified of endpoint sync?

dcbw replied:
@thockin userspace proxy doesn't implement EndpointsHandler, the LB does. So we have to somehow notify the userspace proxy of when the endpoints have been synced.

However, I guess we could just have the userspace proxy implement EndpointsHandler and pass the calls through to the load balancer object that it was originall given? That's a larger code change though.

dcbw replied:
Implemented EndpointsHandler passthrough.

thockin reviewed (commented):
IMO this code is as dead as it could be.  The only significant user is OpenShift as far as I know.  I'd rather never touch it again, but I know that is not realistic.

Also, it seems like maybe this could be broken into a couple commits for easier review?

I raised some questions about this design, but I think you should add yourselves as approvers in OWNERS for this subdir.  If it evolves, I will lose context on the impl.  I don't think it is covered by e2e, either (more argument for breaking it to a separate repo and having its own e2e tests)

dcbw commented:
> Also, it seems like maybe this could be broken into a couple commits for easier review?

@thockin broken into two commits; second commit is largely testcase updates.

dcbw commented:
@thockin PTAL thanks!

danwinship commented on pkg/proxy/userspace/loadbalancer.go:
```diff
@@ -31,4 +31,17 @@ type LoadBalancer interface {
+
+	// OnEndpointsAdd is called whenever creation of new endpoints object
+	// is observed.
+	OnEndpointsAdd(endpoints *v1.Endpoints)
```
Just add `EndpointsHandler` to the top of the interface definition

danwinship commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -92,14 +94,28 @@ func logTimeout(err error) bool {
+const (
+	serviceEventAdd    int = 1
+	serviceEventUpdate int = 2
+	serviceEventDelete int = 3
```
stylistically, you should `type serviceEventType int` or something, and also use `iota`

danwinship commented on pkg/proxy/userspace/proxier_test.go:
```diff
@@ -86,6 +86,17 @@ func waitForClosedPortUDP(p *Proxier, proxyPort int) error {
 }
 
+func waitForServiceInfo(p *Proxier, service proxy.ServicePortName) (*ServiceInfo, bool) {
+	for i := 0; i < 50; i++ {
```
`utilwait.PollImmediate()`?

danwinship commented on pkg/proxy/userspace/proxier_test.go:
```diff
@@ -894,7 +914,17 @@ func TestProxyUpdatePortal(t *testing.T) {
-	_, exists := p.getServiceInfo(service)
+
+	// Wait until the proxy updates the service; fail if it doesn't do so
+	// before the timeout is reached
```
why can't this use `waitForServiceInfo`?

dcbw replied:
@danwinship this one is actually waiting for a serviceInfo that already exists to go away, so it can't use waitForServiceInfo(). I've updated the comment to make that clear.

dcbw commented:
@danwinship maybe also PTAL?

dcbw commented:
/test pull-kubernetes-node-e2e

danwinship commented:
> can't go all the way without converting the Windows userspace proxier though, which should be done, but I'd prefer not to do it right now.

Blah... I'd missed that. I feel like this isn't actually a simplification then and it was better how it was before.

> this one is actually waiting for a serviceInfo that already exists to go away, so it can't use waitForServiceInfo(). I've updated the comment to make that clear.

Yeah, I got that, but I was thinking "it's the same thing, you're just checking for service == nil rather than service != nil at the end", but I missed the fact that you can stop polling once the service no longer exists in this case.

So lgtm except that I feel like probably reverting the `ProxyEventHandler` part would be good?

dcbw commented:
/retest

dcbw commented:
/retest

dcbw commented:
/retest

dcbw commented:
/retest

dcbw commented:
@danwinship PTAL thanks!

dcbw commented:
@danwinship Ok, how about now. I just went all the way with the consolidation.

danwinship commented:
/lgtm

dcbw commented:
@thockin how does this one look now?

thockin commented:
Not kidding - you should add yourselves as OWNERS. :)

thockin commented on cmd/kube-proxy/app/server.go:
```diff
@@ -434,8 +434,6 @@ type ProxyServer struct {
 	OOMScoreAdj            *int32
 	ResourceContainer      string
 	ConfigSyncPeriod       time.Duration
-	ServiceEventHandler    config.ServiceHandler
```
How obnoxious would it be if I asked for this simplification to be a commit of its own?

dcbw replied:
@thockin do you mean just the service event handler distinct from the endpoints? Or something else...

thockin replied:
I meant the removal of service/enpoint handlers as distinct things.  It's a small but clear refactoring.

thockin commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -510,20 +571,70 @@ func (proxier *Proxier) unmergeService(service *v1.Service, existingPorts sets.S
+	proxier.mu.Lock()
+	proxier.serviceEvents = append(proxier.serviceEvents, e)
+	// Sync immediately to reduce memory usage if we get too many events
+	syncNow := len(proxier.serviceEvents) > maxServiceEvents
```
This seems to defeat the optimization of bounded-frequency?  E.g. I could still spam you with meaningless Service changes and pound on iptables..

thockin reviewed (commented):
Overall LGTM.  2 questions

thockin commented:
ping

dcbw commented:
@thockin still working on it...

dcbw commented:
@thockin new approach to address your comments is a lighter-weight version of service changes which instead of stuffing each add/update/del into a list, keeps a map of changed services so that the map will max out at the number of services in your cluster. This will prevent spamming changes as there will only ever be the most recent change in the map. Does this sound better?

dcbw committed:
proxy/userspace: respect minSyncInterval and simplify locking

The userspace proxy does not have any ratelimiting and when many
services are used will hammer iptables every time a service or
endpoint change occurs. Instead build up a map of changed
services and process all those changes at once instead of each
time an event comes in. This also ensures that no long-running
processing happens in the same call chain as the OnService*
calls as this blocks other handlers attached to the proxy's
parent ServiceConfig object for long periods of time.

Locking can also now be simplified as the only accesses to the
proxy's serviceMap happen from syncProxyRules(). So instead of
locking in many functions just lock once in syncProxyRules()
like the other proxies do.

https://bugzilla.redhat.com/show_bug.cgi?id=1590589
https://bugzilla.redhat.com/show_bug.cgi?id=1689690

dcbw commented:
/retest

dcbw committed:
proxy/userspace: add proxy shutdown function and use in testcases

If a testcase does time out and 'go test' prints the call stack,
make sure everything from previous tests is cleaned up so the call
stack is easier to understand.

dcbw committed:
proxy/userspace: track initial service/endpoints sync

We'll use this shortly to prevent premature syncing before all
initial endpoints and services have been received from the
apiserver.

dcbw committed:
proxy/userspace: replace IsServiceIPSet() with ShouldSkipService()

Keeps things consistent with iptables/IPVS proxies. Proxies don't
handle ServiceTypeExternalName even if the ClusterIP is set.

dcbw committed:
proxy/userspace: consolidate portal and proxy cleanup

dcbw commented:
@thockin I've split out as many things into individual commits as I think I can; let me know if you want me to try splitting "proxy: consolidate ServicesHandler/EndpointsHandler into ProxyProvider" though. Otherwise, PTAL thanks!

dcbw commented:
@danwinship could you do another review too? Thanks!

dcbw commented:
@JacobTanenbaum also might be of interest to you since you're looking at the proxy these days.

danwinship commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -287,6 +290,20 @@ func CleanupLeftovers(ipt iptables.Interface) (encounteredError bool) {
 
+// Shutdown closes all service port proxies and returns from the proxy's
+// sync loop
+func (proxier *Proxier) Shutdown() {
```
I'd make this internal (ie, `shutdown()`) and just for testing. Otherwise you have to prove to me that dropping the mutex temporarily partway through the shutdown (between `close(proxier.stopChan)` and `proxier.cleanupStaleStickySessions()`) can't result in race conditions in any possible scenario.

dcbw replied:
The locking-in-the-middle is only used in this interim commit; that gets fixed later when we remove the locking from cleanupStaleStickySessions(). But changed anyway since nothing other than testcsaes uses it.

danwinship commented on pkg/proxy/userspace/proxier_test.go:
```diff
@@ -225,6 +225,15 @@ func waitForNumProxyClients(t *testing.T, s *ServiceInfo, want int, timeout time
 	t.Errorf("expected %d ProxyClients live, got %d", want, got)
 }
 
+func startProxier(p *Proxier, t *testing.T) {
```
This is adding a `SyncLoop` thread and calls to `OnServiceSynced` and `OnEndpointsSynced`, but that has nothing to do with the cleanup/shutdown changes. It just snuck in here from a later commit.

danwinship commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -304,6 +310,26 @@ func (proxier *Proxier) Shutdown() {
 	close(proxier.stopChan)
 }
 
+func (proxier *Proxier) setInitialized(servicesSynced, endpointsSynced bool) {
```
This helper is more confusing than helpful, especially the fact that `setInitialized(true, false)` doesn't mean "set `endpointsSynced` to not initialized".

danwinship commented on pkg/proxy/userspace/proxier.go:
```diff
@@ -557,22 +595,83 @@ func (proxier *Proxier) unmergeService(service *v1.Service, existingPorts sets.S
+	// Must sync from a goroutine to avoid blocking the
+	// service event handler on startup with large numbers
+	// of initial objects
+	go proxier.syncProxyRules()
```
why not just `proxier.syncRunner.Run()`? (which also ensures that it runs from a goroutine, not the calling thread)
likewise in `OnEndpointsAdd`

dcbw replied:
@danwinship because proxier.syncRunner.Run() respects the ratelimiting, which we don't want here. That said until we are initialized we shouldn't ever end up calling syncRunner.Run() so maybe it's OK to call Run() from here right after initializing, I guess.

dcbw replied:
Turns out calling the syncrunner here does have an effect, it doubles the time the testcase takes to run. I'd rather be explicit about what we want here and not rely on the sync runner to "do the right thing" in this case because we know it might not given its purpose in life.

danwinship commented on pkg/proxy/userspace/proxier_test.go:
```diff
@@ -234,6 +246,17 @@ func startProxier(p *Proxier, t *testing.T) {
 }
 
+// addServiceOnPortLocked is only used for testcases
+func addServiceOnPortLocked(p *Proxier, service proxy.ServicePortName, protocol v1.Protocol) (*ServiceInfo, int, error) {
```
returning the port and using that everywhere rather than `svcInfo.proxyPort` just seems like a lot of code churn for no real advantage?

danwinship reviewed (commented):
typo in the last commit message ("consoldiate")

dcbw commented:
@danwinship @thockin I believe I've addressed all comments. PTAL thanks!

dcbw commented:
/retest

danwinship commented:
/lgtm

dcbw commented:
@thockin PTAL thanks!

thockin reviewed (commented):
/lgtm
/approve

dcbw commented:
@thockin thanks Tim!
